test/shared/src/main/scala/zio/test/Gen.scala

```diff
@@ -18,47 +18,99 @@ package zio.test
 
 import zio.{ UIO, ZIO }
 import zio.random._
-import zio.stream.ZStream
+import zio.stream.{ Stream, ZStream }
 
 /**
  * A `Gen[R, A]` represents a generator of values of type `A`, which requires
  * an environment `R`. Generators may be random or deterministic.
  */
 case class Gen[-R, +A](sample: ZStream[R, Nothing, Sample[R, A]]) { self =>
-  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self zip that
 
-  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] =
-    Gen(sample.flatMap(sample => f(sample.value).sample))
+  final def <*>[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] =
+    self zip that
 
-  final def map[B](f: A => B): Gen[R, B] = Gen(sample.map(_.map(f)))
+  final def filter(f: A => Boolean): Gen[R, A] =
+    flatMap(a => if (f(a)) Gen.const(a) else Gen.empty)
 
-  final def zip[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] = self.flatMap(a => that.map(b => (a, b)))
+  final def flatMap[R1 <: R, B](f: A => Gen[R1, B]): Gen[R1, B] = Gen {
+    sample.flatMap { a =>
+      f(a.value).sample.map { b =>
+        val rest = a.shrink.flatMap(a => f(a).sample).flatMap(sample => ZStream(sample.value) ++ sample.shrink)
+        Sample(b.value, b.shrink ++ rest)
+      }
+    }
+  }
+
+  final def map[B](f: A => B): Gen[R, B] =
+    Gen(sample.map(_.map(f)))
+
+  final def noShrink: Gen[R, A] =
+    Gen(sample.map(sample => Sample.noShrink(sample.value)))
+
+  final def withShrink[R1 <: R, A1 >: A](f: A => ZStream[R1, Nothing, A1]): Gen[R1, A1] =
+    Gen(sample.map(sample => Sample(sample.value, f(sample.value) ++ sample.shrink.flatMap(f))))
+
+  final def zip[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (A, B)] =
+    self.flatMap(a => that.map(b => (a, b)))
+
+  final def zipAll[R1 <: R, B](that: Gen[R1, B]): Gen[R1, (Option[A], Option[B])] =
+    Gen {
+      (self.sample.map(Some(_)) ++ Stream(None).forever)
+        .zip(that.sample.map(Some(_)) ++ Stream(None).forever)
+        .collectWhile {
+          case (Some(a), Some(b)) => a.zipWith(b) { case (a, b) => (Some(a), Some(b)) }
+          case (Some(a), None)    => a.map(a => (Some(a), None))
+          case (None, Some(b))    => b.map(b => (None, Some(b)))
+        }
+    }
 
   final def zipWith[R1 <: R, B, C](that: Gen[R1, B])(f: (A, B) => C): Gen[R1, C] =
     (self zip that) map f.tupled
 }
+
 object Gen {
 
+  /**
+   * A generator of booleans.
+   */
+  final val boolean: Gen[Random, Boolean] =
+    choose(false, true)
+
+  final def char(range: Range): Gen[Random, Char] =
```

Comment:
Would be useful to add `identifierChar`, `printableChar`, `unicodeChar`, etc.

Reply:
Yes.